File: sdk/java/driversdk/src/main/java/com/baidu/bce/sdk/plugin/GrpcReport.java
package com.baidu.bce.sdk.plugin;

import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import proto.DriverOuterClass;
import proto.ReportGrpc;

public class GrpcReport implements IReport {
    private final ManagedChannel channel;
    private final ReportGrpc.ReportBlockingStub blockingStub;

    public GrpcReport(String host, int port) {
        this.channel = ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
        this.blockingStub = ReportGrpc.newBlockingStub(channel);
    }

    public void shutdown() {
        channel.shutdown();
    }

    @Override
    public String post(String data) {
        DriverOuterClass.RequestArgs args = DriverOuterClass.RequestArgs.newBuilder().setRequest(data).build();
        DriverOuterClass.ResponseResult response = blockingStub.post(args);
        return response.getData();
    }

    @Override
    public String state(String data) {
        DriverOuterClass.RequestArgs args = DriverOuterClass.RequestArgs.newBuilder().setRequest(data).build();
        DriverOuterClass.ResponseResult response = blockingStub.state(args);
        return response.getData();
    }
}

File: sdk/csharp/driversdk/DriverSDK/Plugin/GrpcReport.cs
using Grpc.Core;
using Proto;

namespace DriverSDK.Plugin
{
	public class GrpcReport: IReport
	{
        private readonly Channel _channel;
        private readonly Report.ReportClient _client;

        public GrpcReport(string host, int port)
        {
            _channel = new Channel(host, port, ChannelCredentials.Insecure);
            _client = new Report.ReportClient(_channel);
        }

        public string Post(string data)
        {
            var request = new RequestArgs { Request = data };
            var response = _client.Post(request);
            return response.Data;
        }

        public string State(string data)
        {
            var request = new RequestArgs { Request = data };
            var response = _client.State(request);
            return response.Data;
        }

        public void Dispose()
        {
            _channel.ShutdownAsync().Wait();
        }

        ~GrpcReport()
        {
            Dispose();
        }
    }
}

